Sort modules by the specified criteria on #306

mitogh opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
The introduced changes creates a predictable outcome of modules by using the criteria specified in #306 to order the modules first by focus area, then by experimental and lastly by module name, in this particular case the slug was used as the name due the slug is normalized into a lowercase strings.

Fixes #306

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

mitogh committed:
Sort modules by the specified criteria on #306

The introduced changes creates a predictable outcome of modules by
using the criteria specified in #306 to order the modules first by
focus area, then by experimental and lastly by module name, in this
particular case the slug was used as the name due the slug is normalized
into a lowercase strings.

Fixes #306

felixarntz commented on bin/plugin/commands/common.js:
```diff
@@ -68,5 +82,26 @@ exports.getModuleData = async ( modulesDir ) => {
+					: -1;
+			}
+
+			const bothModulesAreExperimental =
+				firstModule.experimental && secondModule.experimental;
+			if (
+				! bothModulesAreExperimental &&
+				( firstModule.experimental || secondModule.experimental )
+			) {
```
This looks overly complicated and potentially slightly off? We only need to compare based on experimental state if only one of the two modules to compare is experimental, otherwise this property isn't relevant as they are equal in that sense (similar to how you do it for the focus areas).
```suggestion
			if ( firstModule.experimental !== secondModule.experimental ) {
```

mitogh replied:
Woops looks like I over complicated things on that one, agree just checking on different values would be enough at this level, we might need some additional level inside of the block but that is a different story. :+1:

mitogh replied:
Actually looks like the body of the code should work as it is :)

felixarntz commented on bin/plugin/commands/common.js:
```diff
@@ -33,6 +45,8 @@ exports.getModuleData = async ( modulesDir ) => {
 			const moduleData = {
 				slug: path.basename( moduleDir ),
 				focus: path.basename( path.dirname( moduleDir ) ),
+				// Set default values, in order to allow to avoid errors when ordering, if set property would be override.
+				experimental: false,
```
Why is this needed? Every module needs to have an `Experimental` header, so I think we should remove this default and instead ensure that it is required below, when calling `Array.filter` to remove incomplete entries.

mitogh replied:
Mainly due this property is set dynamic via the headers, in case a header is not set we ensure this property is initialized to a `false` value.

felixarntz replied:
@mitogh See my reply to the other comment. I like your idea there with a warning, let's rely primarily on validation like that. In other words, I still think the default here can be removed as this will be validated later anyway.

felixarntz commented on bin/plugin/commands/common.js:
```diff
@@ -68,5 +82,26 @@ exports.getModuleData = async ( modulesDir ) => {
 			return moduleData;
 		} )
-		.filter( ( moduleData ) => moduleData.name && moduleData.description );
+		.filter( ( moduleData ) => moduleData.name && moduleData.description )
```
See my comment above, it would be better to remove the above default and instead do the following extra validation here:
```suggestion
		.filter( ( moduleData ) => moduleData.name && moduleData.description && typeof moduleData.experimental !== undefined )
```

mitogh replied:
Agree that this would work as well, however I see it like we have 2 options:

1. Initialize to a safe default in case the experimental value is missing in order to include a module.
2. Not include the file but create a warning that a file is missing a required property.

Removing a module by missing a header without an error message might introduce some silent errors if we are not being explicit either on the parameters or error message on the reason why a module is not included. 

Let me know what do you think about that :)

felixarntz replied:
@mitogh That's a good point. How about we expand the `filter` function to also trigger warnings? I think it's okay to not error, we can just ignore such modules technically, but at least there should be a corresponding console output with warnings.

In other words, I think we should go with my originally suggested idea in terms of conditions to check, but expand the function based on your idea that if any of the 3 conditions is not satisfied, there should be a warning triggered in the console. Probably most helpful would be to have a specific warning per which of the 3 checks went wrong.

mitogh replied:
Great I think that's a good call, let me update this filter to include those warnings.

felixarntz reviewed (commented):
@mitogh Looks great, thanks! Just a few minor things worth improving.

mitogh committed:
Simplify conditional for module sorting

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

mitogh committed:
Update filter to remove experimental modules

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

mitogh committed:
Add warnings when a module is missing a required property

Add warngins back to the terminal in the scenario where a module is
lacking a set of the required properties.

Additionally fixed the fact that the `undefined` check was done against
the object and would result in false values.

felixarntz reviewed (approved):
Great work @mitogh!

felixarntz commented:
@eugene-manuilov Can you please review this so that we can merge it soon?

eugene-manuilov commented on bin/plugin/commands/common.js:
```diff
@@ -68,5 +81,52 @@ exports.getModuleData = async ( modulesDir ) => {
-		.filter( ( moduleData ) => moduleData.name && moduleData.description );
+		.filter( ( moduleData ) => {
+			const moduleDetails = JSON.stringify( moduleData );
+			if ( ! moduleData.name ) {
+				log(
+					formats.warning(
+						`This module was not included due is missing a required property 'name', Details: ${ moduleDetails }`
+					)
+				);
+			}
+
+			if ( ! moduleData.description ) {
+				log(
+					formats.warning(
+						`This module was not included due is missing a required property 'description', Details: ${ moduleDetails }`
+					)
+				);
+			}
+
+			if ( typeof moduleData.experimental === 'undefined' ) {
+				log(
+					formats.warning(
+						`This module was not included due is missing a required property 'experimental', Details: ${ moduleDetails }`
+					)
+				);
+			}
```
Please, return `false` early to avoid having multiple warning messages for the same module if it has multiple properties empty. Or maybe would be better to determine which properties are missed and then output a single unified warning message with all missing properties. Something like this:

```
This module was not included because it misses required properties: 'name', 'experimental'.
```

eugene-manuilov commented on bin/plugin/commands/common.js:
```diff
@@ -68,5 +81,52 @@ exports.getModuleData = async ( modulesDir ) => {
+		.sort( ( firstModule, secondModule ) => {
+			// Not the same focus group.
+			if ( firstModule.focus !== secondModule.focus ) {
+				return FOCUS_AREAS[ firstModule.focus ] >
+					FOCUS_AREAS[ secondModule.focus ]
+					? 1
+					: -1;
```
```suggestion
				return FOCUS_AREAS[ firstModule.focus ] - FOCUS_AREAS[ secondModule.focus ];
```

eugene-manuilov reviewed (changes requested):
Almost looks good to me. Left a comment and a nit-pick suggestion. Cris, please, take a look.

mitogh commented:
Thanks @eugene-manuilov would address your PR feedback over the weekend.

mitogh committed:
Merge branch 'trunk' into fix/306-order-modules-by-custom-criteria

mitogh committed:
Use substraction to define the order of the modules.

Co-authored-by: Eugene Manuilov <manuilov@google.com>

mitogh committed:
Merge branch 'fix/306-order-modules-by-custom-criteria' of github.com:WordPress/performance into fix/306-order-modules-by-custom-criteria

mitogh committed:
Combine missing properties into a single error.

Based on the feedback a single warning is displayed when
multiple problems exists in the module order.

eugene-manuilov commented on bin/plugin/commands/common.js:
```diff
@@ -68,5 +81,53 @@ exports.getModuleData = async ( modulesDir ) => {
+					.join( ', ' );
+				log(
+					formats.warning(
+						`This module was not included because it misses required properties: ${ properties }. \nDetails: ${ JSON.stringify(
+							moduleData
+						) }`
```
Let's probably output formatted details:
```suggestion
						`This module was not included because it misses required properties: ${ properties }.\nDetails: ${ JSON.stringify( moduleData, null, 4 ) }`
```

eugene-manuilov reviewed (approved):
Looks good to me. Thanks, @mitogh. Added one more suggestion that is just a nitpick and shouldn't block this PR. Approved.

mitogh committed:
Remove empty space

Co-authored-by: Eugene Manuilov <manuilov@google.com>

felixarntz commented:
@eugene-manuilov @mitogh Can one of you please address the JS lint error here?

**Update:** Fixed it below.

Commit:
Fix JS lint error.
